### ISSUE TITLE
Comment on first line of shader program causes an inaccurate/non-intuitive error message

### ISSUE THREAD

**Issue by ArdaE** [godotengine/godot#16032]
If I place a comment at the top of a shader program, it generates an unrelated/inaccurate "Assignment to constant expression" error message later in the program, e.g. on a line where I assign a value to COLOR. Here's a very simple code snippet to reproduce the issue:

```
// My test shader
shader_type canvas_item;

void fragment( )
{
	COLOR = vec4(1.0,1.0,1.0,1.0);
}
```

Delete the first line, and everything works fine. With the first line there, the error message is displayed on the line that assigns to COLOR. I think comments should be allowed before shader_type is defined, but if the core developers think otherwise, a more accurate error message would also be acceptable. Currently the error message is confusing.

**Godot version:**
V3.0, custom build (currently based on Commit #44cce2b of Jan 24, 2018, but the issue has been there for awhile).

**Comment by brainsick** [comment]
As far as I can tell, this is a discrepancy in how `ShaderLanguage::get_shader_type` and `ShaderLanguage::_parse_shader` detect the shader_type.  `_parse_shader` skips over comments and `get_shader_type` does not.  This causes the compile driver/caller to setup the wrong environment; it defaults to spatial which doesn't allow a fragment shader to write to COLOR, hence the subsequent error.

I'm still poking at this.

**Comment by mosujiba** [comment]
I can confirm that this bug is still present in v3.0.2. The editor will also throw an error if you try to assign a value to LIGHT.

```
// Just a harmless comment, right?

shader_type canvas_item;

void light() {
    LIGHT = vec4(0.0); // error: assignment to constant expression
}
```

**Comment by akien-mga** [comment]
Fixed by #16052 / 9423f23ff.
